Repopulate a StepByStepPage with content from publishing-api

leenagupte opened the pull request:
Trello: https://trello.com/c/c4z7hAxs

Dependant on PR #487

## Motivation

When draft changes are made to a published step by step in collections publisher, there is no way to discard those draft changes without manually unpicking them.

However, all of the previous editions are being stored in Publishing API.

We should give publishers the ability to discard a draft in collections-publisher by getting the latest published version of the step by step from the publishing-api and using that to repopulate the data for the step by step in collections-publisher.

In other words, programmatically unpicking the changes so that the publisher doesn't have to.

## What's changed

This is the first step in the ability to overwrite step by step data in collections-publisher with a version of the step by step held by publishing-api. The service takes any step by step payload from publishing-api and uses it to update the step by step in the database.

sihugh commented on app/services/step_by_step_page_reverter.rb:
```diff
@@ -52,4 +53,12 @@ def steps
+      content[:text]
+    end
+
+    contents_list.join("\r\n")
```
I'm interested in this new line.  I guess it works?  Does `\n` work too?

leenagupte replied:
@sihugh I looked at how steps are stored in the database and just copied that!

leenagupte replied:
Actually @sihugh, I think I need to join on a double newline, so that paragraphs and lists have an empty line between them.

sihugh commented on spec/services/step_by_step_page_reverter_spec.rb:
```diff
@@ -83,6 +83,12 @@
+
+      describe "#contents" do
+        it "saves the contents of a paragraph" do
+          expect(step_by_step_page.steps[0].contents).to eq("A paragraph of text in the first step.")
```
Can we test for the multiline thing?

leenagupte replied:
I have a test for a multi-line, multi-list, multi-bullets, with and without context monstrosity of a step in a later commit.

leenagupte replied:
@sihugh I've updated the test to store multiple paragraphs. I felt it was appropriate as I mentioned it in the comments. Cheers!

sihugh commented on app/models/step.rb:
```diff
@@ -2,7 +2,7 @@ class Step < ApplicationRecord
   validates_presence_of :step_by_step_page
   validates :title, :logic, presence: true
-  has_many :link_reports
+  has_many :link_reports, :dependent => :destroy
```
Same comment as on the other PR - should we be explicit about this in a model test?

leenagupte replied:
I cherry-picked in the commit from the other PR to get the tests to work. I'll unpick it when the other PR is done/merged.

leenagupte commented on app/services/step_by_step_page_reverter.rb:
```diff
@@ -11,6 +11,7 @@ def repopulate_from_publishing_api
     step_by_step_page.slug = payload_from_publishing_api[:base_path].tr('/', '')
     step_by_step_page.introduction = introduction
     step_by_step_page.description = payload_from_publishing_api[:description]
+    step_by_step_page.updated_at = public_updated_at
```
Hmmm. Perhaps not. I'll take it out.

leenagupte replied:
I'll change the test to match the one for `created_at`.

leenagupte replied:
Actually, I think you're @sihugh and it's not needed at all. `updated_at` is updated every time the step nav is saved, so the value will have changed again when the publishers are ready to save and push to publishing-api.

leenagupte commented on app/services/step_by_step_page_reverter.rb:
```diff
@@ -12,6 +12,7 @@ def repopulate_from_publishing_api
     step_by_step_page.introduction = introduction
     step_by_step_page.description = payload_from_publishing_api[:description]
     step_by_step_page.updated_at = public_updated_at
+    step_by_step_page.published_at = public_updated_at
```
I think you're right. I was being overly cautious. I'll take it out. Cheers!

leenagupte replied:
I'll add a test to check it isn't overwritten ala `created_at`

leenagupte commented:
@sihugh I've addressed your comments and fixed up. Github is showing the commits in an odd order though.

They should be ordered thusly:
```
Repopulate title of step nav from publishing-api
Repopulate the slug of the step by step
Repopulate the introduction of the step by step
Repopulate the description of a step by step
Don’t repopulate the content-id for a step by step
Don’t repopulate the created date of the step by step
Don't repopulate the published_at date of the step by step
Repopulate draft_updated_at time with public_updated_at
Populate the logic for each step
Populate if the steps in the step by step are optional
Populate the position of each step
Saves the contents of a step that is just text
Saves the contents of a step that contains links
Saves contents of step with bulleted list of links
Saves contents of step with bulleted list of text
Saves lists of links with context
Saves bulleted lists of links with context
Check content can save links and paragraphs
Save Navigation Rules for all internal links
Turn off step nav for rules in `pages_related_to_step_nav`
```

leenagupte commented:
@sihugh It's the next story!

sihugh commented:
MAYBE JUST RUN IT AS A RAKE TASK AT RANDOM INTERVALS 👿

leenagupte commented:
@sihugh :trollface:
